### PR TITLE
Identity add brokered auth

### LINKED ISSUE #27891: Add Windows Web Account Manager (WAM) Brokered Authentication Support
This is a tracking work item for creating/finishing the new Azure.Identity.BrokeredAuthentication client library (separate package from the existing Azure Identity client library).

Related:
- https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity.BrokeredAuthentication/README.md

Dependencies:
- MSAL WAM is supported (Beta) in .NET and Python today
- MSAL WAM support (Beta) for Java and Javascript is not yet available but is expected to be available in time for this scheduled development milestone.

### PR DESCRIPTION
**Pull request by billwert**
This enables using the Web Account Manager on Windows.

Adds a new library (`azure-identity-broker`) and surface area in `azure-identity` to enable the feature.

Fixes #27891

### CODE REVIEW

File: sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/InteractiveBrowserBrokerCredentialBuilder.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+package com.azure.identity.broker;
+
+
+import com.azure.identity.InteractiveBrowserCredential;
+import com.azure.identity.InteractiveBrowserCredentialBuilder;
+
+/**
+ * Fluent credential builder for instantiating a {@link InteractiveBrowserCredential} configured to use a secure broker.
+ *
```
Comment (g2vinay) [inline sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/InteractiveBrowserBrokerCredentialBuilder.java:12]: Docs to add code samples to use the builder.

File: sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/implementation/package-info.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+
```
Comment (g2vinay) [inline sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/implementation/package-info.java:3]: Docs: This needs to be updated to have detailed docs. azure-identity can be used as reference.

File: sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/package-info.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/**
+ * <p>The Azure Identity Brokered Authentication library offers support for using various support brokers from the
+ * MSAL4j library.</p>
```
Comment (g2vinay) [inline sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/package-info.java:6]: Docs: This needs to be updated to have detailed docs. azure-identity can be used as reference.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java [hunk 4]
```diff
@@ -318,6 +324,37 @@ PublicClientApplication getPublicClient(boolean sharedTokenCacheCredential, bool
             builder.clientCapabilities(set);
         }
 
+        if (options.isBrokerEnabled()) {
+            if (interactiveBrowserBroker == null) {
+                try {
+                    interactiveBrowserBroker = Class.forName("com.azure.identity.broker.implementation.InteractiveBrowserBroker");
+                } catch (ClassNotFoundException e) {
+                    throw LOGGER.logExceptionAsError(new RuntimeException("Could not load the brokered authentication library. "
+                        + "Please ensure that the azure-identity-broker library is on the classpath.", e));
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:333]: Please prefix isn't needed in error messages. :)

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java [hunk 5]
```diff
@@ -318,6 +324,37 @@ PublicClientApplication getPublicClient(boolean sharedTokenCacheCredential, bool
...
+                }
+                getMsalRuntimeBroker = null;
+                try {
+                    getMsalRuntimeBroker = interactiveBrowserBroker.getMethod("getMsalRuntimeBroker");
+                } catch (NoSuchMethodException e) {
+                    throw LOGGER.logExceptionAsError(new RuntimeException("Could not obtain the InteractiveBrowserBroker. "
```
[the remark this answers is not in the document]
Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:339]: What I really wanted was `PlatformNotSupportedException` from C#. Maybe `IllegalStateException` is the best alternative? `UnsupportedOperationException` is also a candidate.

[the remark this answers is not in the document]
Comment (xiangyan99) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:339]: Should we raise something like `NotImplementedError`?  Maybe `UnsupportedOperationException` for Java?

**Comment by billwert** [comment]
RE: Doc fixes. Please leave the comments. I will address them for GA.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:339]: Made it `IllegalStateException` for now. `UnsupportedOperationException` says it is part of the Collections framework so I bet the semantic isn't quite what folks would expect here. Open to better ideas!

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java [hunk 6]
```diff
@@ -318,6 +324,37 @@ PublicClientApplication getPublicClient(boolean sharedTokenCacheCredential, bool
...
+                        + "Please ensure that the azure-identity-broker library is on the classpath.", e));
+                }
+            }
+
+            try {
+                if (getMsalRuntimeBroker != null) {
+                    builder.broker((IBroker) getMsalRuntimeBroker.invoke(null));
+
+                } else {
+                    throw LOGGER.logExceptionAsError(new RuntimeException("Could not obtain the MSAL Broker. "
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:349]: Runtime -> IllegalStateException ?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java [hunk 7]
```diff
@@ -318,6 +324,37 @@ PublicClientApplication getPublicClient(boolean sharedTokenCacheCredential, bool
...
+                        + "Please ensure that the azure-identity-broker library is on the classpath.", null));
+                }
+            } catch (InvocationTargetException | IllegalAccessException e) {
+                throw LOGGER.logExceptionAsError(new RuntimeException("Could not invoke the MSAL Broker. "
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientBase.java:353]: Runtime -> IllegalStateException ?

File: sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/InteractiveBrowserBrokerCredentialBuilder.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+     * @param windowHandle The window handle of the current application, or 0 for a console application.
+     * @return An updated instance of this builder with the interactive browser broker configured.
+     */
+    public InteractiveBrowserBrokerCredentialBuilder setWindowHandle(long windowHandle) {
+        this.identityClientOptions.setBrokerWindowHandle(windowHandle);
+        return this;
+    }
+
+    /**
+     * Enables Microsoft Account (MSA) passthrough.
```
Comment (g2vinay) [inline sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/InteractiveBrowserBrokerCredentialBuilder.java:27]: Docs To Add : What happens by enabling it ?

File: sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/InteractiveBrowserBrokerCredentialBuilder.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import com.azure.identity.InteractiveBrowserCredentialBuilder;
+
+/**
+ * Fluent credential builder for instantiating a {@link InteractiveBrowserCredential} configured to use a secure broker.
+ *
+ */
+public class InteractiveBrowserBrokerCredentialBuilder extends InteractiveBrowserCredentialBuilder {
+
+    /**
+     * For Windows, sets the parent window handle used by the broker.
```
Comment (g2vinay) [inline sdk/identity/azure-identity-broker/src/main/java/com/azure/identity/broker/InteractiveBrowserBrokerCredentialBuilder.java:17]: For Windows Platform/OS

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 10]
```diff
@@ -76,6 +76,10 @@ public final class IdentityClientOptions implements Cloneable {
 
     private boolean isChained;
     private boolean enableUnsafeSupportLogging;
+    private long brokerWindowHandle;
+    private boolean brokerEnabled;
+    private boolean enableMsaPassthrough;
+    private boolean brokerSupportsWindows;
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:82]: I don't see this variable brokerSupportsWindows being referenced in this class.

File: sdk/identity/azure-identity-broker/CHANGELOG.md [hunk 11]
```diff
@@ -0,0 +1,6 @@
+# Release History
+
+## 1.0.0-beta.1 (Unreleased)
+
+### Features Added
+- Initial release of WAM support for Azure Identity.
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/CHANGELOG.md:6]: Let's assume most customers are unfamiliar with the WAM acronym. 
```suggestion
- Initial release of WAM (Windows authentication broker) support for Azure Identity.
```

File: sdk/identity/azure-identity-broker/README.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
+# Azure Identity Brokered Authentication client library for Java
+
+The Azure Identity Brokered Authentication library extends the Azure Identity library to provide brokdered authentication support. This options class can be used to create an `InteractiveBrowserCredential` capable of using the system authentication broker in lieu of the web browser when available.
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:3]:
```suggestion
The Azure Identity Brokered Authentication library extends the Azure Identity library to provide brokered authentication support. This options class can be used to create an `InteractiveBrowserCredential` capable of using the system authentication broker in lieu of the web browser when available.
```

File: sdk/identity/azure-identity-broker/README.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+- A [Java Development Kit (JDK)][jdk_link], version 8 or later.
+- An [Azure subscription][azure_sub].
+- The Azure CLI can also be useful for authenticating in a development environment, creating accounts, and managing account roles.
+
+
+## Key concepts
+
+### Examples
+
+#### Configuring the `InteractiveBrowserCredential` to use the system authentication broker
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:65]: Avoid gerunds in headings when possible.
```suggestion
#### Configure the `InteractiveBrowserCredential` to use the system authentication broker
```

File: sdk/identity/azure-identity-broker/README.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+```java
+long windowHandle = GetWindowHandle() // Samples below
+InteractiveBrowserCredential cred = new InteractiveBrowserCredentialBuilder()
+    .useInteractiveBrowserBroker(windowHandle)
+    .build();
+```
+
+#### Obtaining a window handle
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:74]:
```suggestion
#### Obtain a window handle
```

File: sdk/identity/azure-identity-broker/README.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+##### JavaFX
+
+This code will obtain the window handle for a JavaFX application for use with `InteractiveBrowserCredentialBuilder`
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:78]:
```suggestion
This code will obtain the window handle for a JavaFX application for use with `InteractiveBrowserCredentialBuilder`:
```

File: sdk/identity/azure-identity-broker/README.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    } catch (Exception e) {
+        e.printStackTrace();
+        return 0;
+    }
+}
+```
+
+## Troubleshooting
+
+Credentials raise exceptions when they fail to authenticate or can't execute authentication. When credentials fail to authenticate, the`ClientAuthenticationException` is raised. The exception has a `message` attribute, which describes why authentication failed. When this exception is raised by `ChainedTokenCredential`, the chained execution of underlying list of credentials is stopped.
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:99]:
```suggestion
Credentials raise exceptions when they fail to authenticate or can't execute authentication. When credentials fail to authenticate, the`ClientAuthenticationException` is raised. The exception has a `message` attribute, which describes why authentication failed. When this exception is raised by `ChainedTokenCredential`, the chained execution of underlying list of credentials is stopped.
```
```suggestion
Credentials raise exceptions when they fail to authenticate or can't execute authentication. When credentials fail to authenticate, the `ClientAuthenticationException` is raised. The exception has a `message` attribute, which describes why authentication failed. When this exception is raised by `ChainedTokenCredential`, the chained execution of underlying list of credentials is stopped.
```

File: sdk/identity/azure-identity-broker/README.md [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+
+When credentials can't execute authentication due to one of the underlying resources required by the credential being unavailable on the machine, the`CredentialUnavailableException` is raised. The exception has a `message` attribute that describes why the credential is unavailable for authentication execution. When this exception is raised by `ChainedTokenCredential`, the message collects error messages from each credential in the chain.
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:101]:
```suggestion
When credentials can't execute authentication due to one of the underlying resources required by the credential being unavailable on the machine, the `CredentialUnavailableException` is raised. The exception has a `message` attribute that describes why the credential is unavailable for authentication execution. When this exception is raised by `ChainedTokenCredential`, the message collects error messages from each credential in the chain.
```

File: sdk/identity/azure-identity-broker/README.md [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+See the [troubleshooting guide](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/identity/azure-identity/TROUBLESHOOTING.md) for details on how to diagnose various failure scenarios.
+
+## Next steps
+
+The Java client libraries listed [here](https://azure.github.io/azure-sdk/releases/latest/java.html) support authenticating with `TokenCredential` and the Azure Identity library. You can learn more about their use, and find additional documentation on use of these client libraries along samples with can be found in the links mentioned [here](https://azure.github.io/azure-sdk/releases/latest/java.html).
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:107]: When the same content exists on Learn, prefer that link instead.
```suggestion
[These Java client libraries](https://learn.microsoft.com/azure/developer/java/sdk/azure-sdk-library-package-index#libraries-using-azure-core) support authenticating with `TokenCredential` and the Azure Identity library. To learn more about using these client libraries, see the aforementioned link.
```

File: sdk/identity/azure-identity-broker/README.md [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+
+When you submit a pull request, a CLA-bot will automatically determine whether you need to provide a CLA and decorate the PR appropriately (e.g., label, comment). Simply follow the instructions provided by the bot. You will only need to do this once across all repos using our CLA.
+
+This project has adopted the [Microsoft Open Source Code of Conduct][code_of_conduct]. For more information see the Code of Conduct FAQ or contact opencode@microsoft.com with any additional questions or comments.
+
+
+```
+<!-- LINKS -->
+[azure_core_library]: https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/core
+[azure_sub]: https://azure.microsoft.com/free/
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:123]:
```suggestion
[azure_sub]: https://azure.microsoft.com/free/java
```

File: sdk/identity/azure-identity-broker/README.md [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+[entra_id_doc]: https://learn.microsoft.com/azure/active-directory/
```
Comment (scottaddie) [inline sdk/identity/azure-identity-broker/README.md:124]:
```suggestion
[entra_id_doc]: https://learn.microsoft.com/entra/identity/
```

File: sdk/identity/azure-identity/CHANGELOG.md [hunk 21]
```diff
@@ -3,12 +3,14 @@
 ## 1.11.0-beta.2 (unreleased)
 
 ### Features Added
+- Initial release of WAM support for Azure Identity.
```
Comment (scottaddie) [inline sdk/identity/azure-identity/CHANGELOG.md:6]:
```suggestion
- Initial release of WAM (Windows authentication broker) support for Azure Identity.
```

**Review by g2vinay — APPROVED** [review]

**Pull request merged**